lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,404 @@
+
+
+BUILD_ASSERT(DATA_BUFFER_SIZE > INPUT_WINDOW_SIZE);
+BUILD_ASSERT(INPUT_WINDOW_SIZE % INPUT_FRAME_SIZE == 0);
+
+
+static size_t buf_calc_max_processing_move(struct data_buffer *b)
+{
+	size_t res;
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->append_head >= b->processing_head) {
+		res = b->append_head - b->processing_head;
+	} else {
+		res = (buf_end - b->processing_head) +
+		      (b->append_head - buf_start);
+	}
+
+	return res;
+}
+
+static size_t buf_calc_free_space(struct data_buffer *b)
+{
+	size_t res;
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	if (b->wait_data_size > 0) {
+		res = b->wait_data_size + ARRAY_SIZE(b->buf) -
+		      INPUT_WINDOW_SIZE - 1;
+	} else {
+		if (b->processing_head > b->append_head) {
+			res = b->processing_head - b->append_head - 1;
+		} else {
+			res = (buf_end - b->append_head) +
+			      (b->processing_head - buf_start) - 1;
+		}
+	}
+
+	return res;
+}
+
+static void buf_processing_end(struct data_buffer *b)
+{
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	__ASSERT_NO_MSG(b->state == STATE_PROCESSING);
+	b->state = STATE_READY;
+
+	k_spin_unlock(&b->lock, key);
+}
+
+static int buf_cleanup(struct data_buffer *b)
+{
+	int err = 0;
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (b->state == STATE_PROCESSING) {
+		err = -EBUSY;
+	} else {
+		b->processing_head = b->buf;
+		b->append_head = b->buf;
+		b->wait_data_size = 0;
+		b->state = STATE_READY;
+	}
+
+	k_spin_unlock(&b->lock, key);
+
+	return err;
+}
+
+static int buf_append(struct data_buffer *b, const float *data, size_t len)
+{
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (buf_calc_free_space(b) < len) {
+		k_spin_unlock(&b->lock, key);
+		return -ENOMEM;
+	}
+
+	float *cur_head = b->append_head;
+	float *new_head = b->append_head + len;
+	int result = 0;
+	bool looped = false;
+
+	if (b->wait_data_size > 0) {
+		if (b->wait_data_size > len) {
+			b->wait_data_size -= len;
+		} else {
+			b->wait_data_size = 0;
+			result = PROCESS_BUF;
+		}
+	}
+
+	if (new_head >= buf_end) {
+		new_head -= ARRAY_SIZE(b->buf);
+		looped = true;
+	}
+
+	b->append_head = new_head;
+
+	k_spin_unlock(&b->lock, key);
+
+	if (looped) {
+		size_t buf_left = buf_end - cur_head;
+
+		memcpy(cur_head, data, buf_left * sizeof(float));
+		memcpy(buf_start, data + buf_left,
+		       (len - buf_left) * sizeof(float));
+	} else {
+		memcpy(cur_head, data, len * sizeof(float));
+	}
+
+	return result;
+}
+
+static void buf_get(struct data_buffer *b, float *b_res, size_t offset,
+		    size_t len)
+{
+	__ASSERT_NO_MSG((offset + len) <= INPUT_WINDOW_SIZE);
+
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	__ASSERT_NO_MSG(b->state == STATE_PROCESSING);
+
+	float *read_start = b->processing_head + offset;
```

Comment:
It would be good to add a comment here that processing head cannot change from any other context while processing is done

Reply:
In such case, considering how this function is used, should there be a spinlock?

Reply:
Comment added.
I removed the spinlock too. I added it before for consistency - accessing all the data related to buffer was done under spinlock.